parser/src/ast/boundary_constraints.rs

```diff
@@ -61,9 +61,14 @@ impl Display for Boundary {
 #[derive(Debug, PartialEq, Clone)]
 pub enum BoundaryExpr {
     Const(u64),
-    /// Reference to a public input element, identified by the name of a public input array and the
-    /// index of the cell.
-    PubInput(Identifier, usize),
+    /// Represents any named constant or variable.
+    Var(Identifier),
+    /// Represents an element inside a constant or variable vector. The index is the index of
+    /// this value inside the vector.
+    VecElem(Identifier, usize),
+    /// Represents an element inside a constant or variable matrix. Indices idx_row and idx_col
+    /// are the indices of this value inside the matrix.
+    MatrixElem(Identifier, usize, usize),
```

Comment:
nit: as mentioned before, I think we should create a new atomic type and use it both here and in the transition constraints (and for grouped columns in the future as well)